File: apps/api/models/score.py
from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel


class ScoreLabel(str, Enum):
    EXCEPTIONAL = "Exceptional"
    STRONG = "Strong"
    SOLID = "Solid"
    MARGINAL = "Marginal"
    WEAK = "Weak"


class ScoreLabelThresholds:
    EXCEPTIONAL = 85.0
    STRONG = 70.0
    SOLID = 55.0
    MARGINAL = 40.0

    @classmethod
    def from_score(cls, score: float) -> ScoreLabel:
        if score >= cls.EXCEPTIONAL:
            return ScoreLabel.EXCEPTIONAL
        elif score >= cls.STRONG:
            return ScoreLabel.STRONG
        elif score >= cls.SOLID:
            return ScoreLabel.SOLID
        elif score >= cls.MARGINAL:
            return ScoreLabel.MARGINAL
        else:
            return ScoreLabel.WEAK


class SubIndicatorScore(BaseModel):
    raw_value: Optional[float] = None
    normalized_score: float
    points_contribution: float
    weight: float


class BonusModifier(BaseModel):
    name: str
    points: float
    reason: str


class ImprovementLever(BaseModel):
    indicator: str
    current_pts: float
    max_pts: float
    gap: float
    hint: str


class OpportunityScoreDetail(BaseModel):
    id: UUID
    listing_id: UUID
    calculated_at: datetime

    # Raw values for all 10 indicators
    raw_land_cost_per_unit: Optional[float] = None
    raw_rent_growth_pct: Optional[float] = None
    raw_population_growth_pct: Optional[float] = None
    raw_median_hh_income: Optional[float] = None
    raw_employment_growth_pct: Optional[float] = None
    raw_supply_pressure_ratio: Optional[float] = None
    raw_effective_tax_rate: Optional[float] = None
    raw_sfha_pct: Optional[float] = None
    raw_wetland_coverage_pct: Optional[float] = None
    raw_infrastructure_score: Optional[float] = None

    # Normalized scores (0–10 per indicator)
    norm_land_cost_per_unit: float
    norm_rent_growth: float
    norm_population_growth: float
    norm_median_hh_income: float
    norm_employment_growth: float
    norm_new_supply: float
    norm_property_tax_burden: float
    norm_flood_risk: float
    norm_wetland_risk: float
    norm_infrastructure_access: float

    # Weighted point contributions
    pts_land_cost_per_unit: float
    pts_rent_growth: float
    pts_population_growth: float
    pts_median_hh_income: float
    pts_employment_growth: float
    pts_new_supply: float
    pts_property_tax_burden: float
    pts_flood_risk: float
    pts_wetland_risk: float
    pts_infrastructure_access: float

    # Bonuses and composite
    bonus_modifiers: list[BonusModifier] = []
    bonus_points_total: float = 0.0
    base_composite_score: float
    final_score: float
    score_label: ScoreLabel
    improvement_levers: list[ImprovementLever] = []


class ScoreModelVersion(BaseModel):
    id: UUID
    version_number: str
    effective_date: date
    is_active: bool
    weights: dict[str, float]
    breakpoints: dict[str, list[dict]]

File: packages/shared-types/src/score.ts
export type ScoreLabel = "Exceptional" | "Strong" | "Solid" | "Marginal" | "Weak";

export const SCORE_LABEL_THRESHOLDS = {
  EXCEPTIONAL: 85,
  STRONG: 70,
  SOLID: 55,
  MARGINAL: 40,
} as const;

export function getScoreLabel(score: number): ScoreLabel {
  if (score >= SCORE_LABEL_THRESHOLDS.EXCEPTIONAL) return "Exceptional";
  if (score >= SCORE_LABEL_THRESHOLDS.STRONG) return "Strong";
  if (score >= SCORE_LABEL_THRESHOLDS.SOLID) return "Solid";
  if (score >= SCORE_LABEL_THRESHOLDS.MARGINAL) return "Marginal";
  return "Weak";
}

export interface SubIndicatorScore {
  raw_value: number | null;
  normalized_score: number;
  points_contribution: number;
  weight: number;
}

export interface BonusModifier {
  name: string;
  points: number;
  reason: string;
}

export interface ImprovementLever {
  indicator: string;
  current_pts: number;
  max_pts: number;
  gap: number;
  hint: string;
}

export interface OpportunityScoreDetail {
  id: string;
  listing_id: string;
  calculated_at: string;

  raw_land_cost_per_unit: number | null;
  raw_rent_growth_pct: number | null;
  raw_population_growth_pct: number | null;
  raw_median_hh_income: number | null;
  raw_employment_growth_pct: number | null;
  raw_supply_pressure_ratio: number | null;
  raw_effective_tax_rate: number | null;
  raw_sfha_pct: number | null;
  raw_wetland_coverage_pct: number | null;
  raw_infrastructure_score: number | null;

  norm_land_cost_per_unit: number;
  norm_rent_growth: number;
  norm_population_growth: number;
  norm_median_hh_income: number;
  norm_employment_growth: number;
  norm_new_supply: number;
  norm_property_tax_burden: number;
  norm_flood_risk: number;
  norm_wetland_risk: number;
  norm_infrastructure_access: number;

  pts_land_cost_per_unit: number;
  pts_rent_growth: number;
  pts_population_growth: number;
  pts_median_hh_income: number;
  pts_employment_growth: number;
  pts_new_supply: number;
  pts_property_tax_burden: number;
  pts_flood_risk: number;
  pts_wetland_risk: number;
  pts_infrastructure_access: number;

  bonus_modifiers: BonusModifier[];
  bonus_points_total: number;
  base_composite_score: number;
  final_score: number;
  score_label: ScoreLabel;
  improvement_levers: ImprovementLever[];
}
